common/djangoapps/student/models.py

```diff
@@ -165,14 +168,39 @@ def anonymous_id_for_user(user, course_id, save=True):
     cached_id = getattr(user, '_anonymous_id', {}).get(course_id)
     if cached_id is not None:
         return cached_id
+    # check if an annonymous id already exists for this user and course_id combination
+    anonymous_user_ids = AnonymousUserId.objects.filter(user=user).filter(course_id=course_id)
+    if anonymous_user_ids:
+        if len(anonymous_user_ids) == 1:
+            anonymous_user_id = anonymous_user_ids[0].anonymous_user_id
+            if not hasattr(user, '_anonymous_id'):
+                user._anonymous_id = {}  # pylint: disable=protected-access
+            user._anonymous_id[course_id] = anonymous_user_id  # pylint: disable=protected-access
+
+            return anonymous_user_id
```

Comment:
Note from discussion: Since this handles the get case, the get_or_create query farther down can be just a create.